src/components/Table/RowState.ts

```diff
@@ -45,6 +45,23 @@ export class RowState {
     // Make ourselves an observable so that mobx will do caching of .collapseIds so
     // that it'll be a stable identity for GridTable to useMemo against.
     makeAutoObservable(this, { rows: false } as any); // as any b/c rows is private, so the mapped type doesn't see it
+
+    // Whenever our `visibleRows` change (i.e. via filtering) then we need to re-derive header and parent rows' selected state.
+    reaction(
+      () => [...this.visibleRows.values()].sort(),
+      () => {
+        const map = new Map<string, SelectedState>();
+        // go through the rows and re-derive all visible parent states.
+        this.rows.current.forEach((row) => this.setNestedSelectedStates(row, map));
+
+        // Finally re-derive the 'header's state.
+        map.set("header", deriveParentSelected(this.getVisibleImmediateChildrenSelectedStates(this.rows.current, map)));
```

Comment:
Is this the same as:

```suggestion
        map.set("header", deriveParentSelected(this.rows.current.flatMap((row) => this.setNestedSelectedStates(row, map)));
```

Reply:
Yea! That works. Thanks for the suggestion.